### PR TITLE
On Hold: Add unit tests for ExecutableDeterminationVisitor

### LINKED ISSUE #643: Add unit tests for ExecutableDeterminationVisitor

### PR DESCRIPTION
**Pull request by lbschanno**
Add a suite of tests for the class ExecutableDeterminationVisitor.
Additionally, do some minor refactoring to do the following:
- Improve clarity of code and add documentation.
- Attempt to improve the efficiency of several methods.
- Reduce overall code duplication.

Fixes #643

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java [hunk 1]
```diff
@@ -68,74 +68,126 @@
...
+         */
+        PARTIAL(1, 1, 2),
+        
+        /**
+         * The expression cannot be executed against the index.
+         */
+        NON_EXECUTABLE(3, 2, 1),
+        
+        /**
+         * The executable state of the underlying expression does not matter.
```
[the remark this answers is not in the document]
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java]: Done.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java [hunk 2]
```diff
@@ -594,390 +405,1040 @@ public Object visit(ASTLTNode node, Object data) {
...
+     *            the data
+     * @param states
+     *            the set of states
+     */
+    private void readStatesFromChildren(final JexlNode node, final Object data, final Set<STATE> states) {
+        final int numChildren = node.jjtGetNumChildren();
+        for (int i = 0; i < numChildren; i++) {
+            states.add((STATE) (node.jjtGetChild(i).jjtAccept(this, data)));
+            // Break early if all possibly unique states have already been added.
+            if (states.size() == STATE.size) {
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java]: this is a good idea, but we need a flag to bypass this for debugging

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java [hunk 3]
```diff
@@ -594,390 +405,1040 @@ public Object visit(ASTLTNode node, Object data) {
...
+        
+        if (log.isTraceEnabled()) {
+            log.trace("node:" + PrintingVisitor.formattedQueryString(node));
+            log.trace("states are:" + states);
+        }
+        
+        // If no states were found, or both EXECUTABLE and NON_EXECUTABLE were found, return PARTIAL.
+        final STATE state;
+        if (states.size() == 0 || (states.size() > 2 && states.first() == STATE.EXECUTABLE)) {
+            state = STATE.PARTIAL;
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java]: This behavior isn't different from before, but a STATE.PARTIAL on no states seems wrong

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java [hunk 4]
```diff
@@ -68,74 +68,126 @@
...
+        
+        /**
+         * The expression does not matter for determining executability against the global index.
+         */
+        IGNORABLE(4, 4, 4),
+        
+        /**
+         * The expression is index-only, but cannot be run against the index (negation, delayed prefix).
+         */
+        ERROR(0, 0, 0);
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java]: I see what you were trying to do here for efficiency, but this impacts clarity of the code quite a bit. In this case I think the clarity/maintainability is more important than the efficiency gain.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java (hunk 2 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java]: Done.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java (hunk 3 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java]: Perhaps NON_EXECUTABLE would make more sense?

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java (hunk 4 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExecutableDeterminationVisitor.java]: I re-implemented this in a different way that removes the need to store the priority ordering in the `STATE` enum. Let me know if you think it requires further changes.

**Comment by ivakegg** [comment]
FineAndDandy is in the process of combining a mess of changes into the executablity visitor which may make some of the refactoring OBE.  The test cases are good and should be incorporated into what he ends up with.  My apologies on the delay here....

**Comment by lbschanno** [comment]
@ivakegg  no problem, thanks for the update.

**Pull request closed without merging**